pkg/collector/corechecks/servicediscovery/module/impl_linux.go

```diff
@@ -613,6 +616,87 @@ func (s *discovery) updateServicesCPUStats(services []model.Service) error {
 	return nil
 }
 
+func (s *discovery) enrichContainerData(service *model.Service, containers map[string]*agentPayload.Container, pidToCid map[int]string) {
+	id, ok := pidToCid[service.PID]
+	if !ok {
+		return
+	}
+
+	// The tags we look for service name generation, in their priority order.
+	// The map entries will be filled as we go through the containers tags.
+	tagsPriority := []struct {
+		tagName  string
+		tagValue *string
+	}{
+		{"service", nil},
+		{"app", nil},
+		{"short_image", nil},
+		{"kube_container_name", nil},
+		{"kube_deployment", nil},
+		{"kube_service", nil},
+	}
+
+	service.ContainerID = id
+	log.Debugf("Found container id for %v: %v", service.Name, id)
+
+	// We got the service name from container tags before, no need to do it again.
+	if service.GeneratedNameFromContainerTags {
+		return
+	}
+
+	container, ok := containers[id]
+	if !ok {
+		return
+	}
+
+	var serviceName string
+
+	for _, tag := range container.Tags {
```

Comment:
Can we add some unit tests for this tag parsing logic to check the edge conditions etc? Shouldn't need mocking etc, just a simple unit test. (I think you'll need to break out the inner logic into its own function.)